File: architecture-skills/object-oriented-programming-paradigm/examples/interfaces-and-abstract-classes/interfaces-and-abstract-classes.py
"""Interfaces & abstract classes in Python.

Python has no `interface` keyword: an abc.ABC with only abstract
methods plays the pure-contract role (typing.Protocol is the
structural alternative), while an ABC with implemented methods is an
abstract class. TabularExporter captures the shared line-assembly
logic once; JsonExporter satisfies the contract directly.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass


@dataclass(frozen=True)
class Metric:
    name: str
    value: float


class Exporter(ABC):
    """The interface: a pure contract, no implementation."""

    @property
    @abstractmethod
    def content_type(self) -> str: ...

    @abstractmethod
    def export(self, metrics: list[Metric]) -> str: ...

    def file_name(self) -> str:
        """A "default method": derived purely from the contract itself."""
        return "report." + self.content_type.rsplit("/", 1)[-1]


class TabularExporter(Exporter):
    """Abstract class: implements export() once, defers the varying parts."""

    @abstractmethod
    def header(self) -> str: ...

    @abstractmethod
    def row(self, metric: Metric) -> str: ...

    def export(self, metrics: list[Metric]) -> str:
        lines = [self.header(), *(self.row(m) for m in metrics)]
        lines.append(f"# {len(metrics)} records")
        return "\n".join(lines)


class CsvExporter(TabularExporter):
    content_type = "text/csv"

    def header(self) -> str:
        return "name,value"

    def row(self, metric: Metric) -> str:
        return f"{metric.name},{metric.value:.2f}"


class JsonExporter(Exporter):
    """Implements the contract directly — no base-class logic needed."""

    content_type = "application/json"

    def export(self, metrics: list[Metric]) -> str:
        entries = ", ".join(
            f'{{"name": "{m.name}", "value": {m.value:.2f}}}' for m in metrics
        )
        return f"[{entries}]"


def deliver(exporter: Exporter, metrics: list[Metric]) -> None:
    """The client programs to the contract, never to a concrete class."""
    print(f"--- {exporter.content_type} -> {exporter.file_name()} ---")
    print(exporter.export(metrics))
    print()


if __name__ == "__main__":
    metrics = [Metric("cpu", 0.93), Metric("memory", 0.67), Metric("disk", 0.41)]
    for exporter in (CsvExporter(), JsonExporter()):
        deliver(exporter, metrics)

File: architecture-skills/object-oriented-programming-paradigm/examples/interfaces-and-abstract-classes/interfaces-and-abstract-classes.js
/** Interfaces & abstract classes in JavaScript.
 *
 * JavaScript has no interface keyword: a class with only method stubs
 * (throwing) or duck typing plays the contract role. TabularExporter
 * implements exportAll once; JsonExporter satisfies the contract directly.
 */

class Metric {
  constructor(name, value) {
    this.name = name;
    this.value = value;
  }
}

class Exporter {
  get contentType() {
    throw new Error('not implemented');
  }

  exportAll(metrics) {
    throw new Error('not implemented');
  }

  fileName() {
    return 'report.' + this.contentType.split('/').pop();
  }
}

class TabularExporter extends Exporter {
  exportAll(metrics) {
    const lines = [this.header(), ...metrics.map((m) => this.row(m))];
    lines.push(`# ${metrics.length} records`);
    return lines.join('\n');
  }

  header() {
    throw new Error('not implemented');
  }

  row(metric) {
    throw new Error('not implemented');
  }
}

class CsvExporter extends TabularExporter {
  get contentType() { return 'text/csv'; }
  header() { return 'name,value'; }
  row(m) { return `${m.name},${m.value.toFixed(2)}`; }
}

class JsonExporter extends Exporter {
  get contentType() { return 'application/json'; }

  exportAll(metrics) {
    const entries = metrics.map(
      (m) => `{"name": "${m.name}", "value": ${m.value.toFixed(2)}}`
    );
    return `[${entries.join(', ')}]`;
  }
}

function deliver(exporter, metrics) {
  console.log(`--- ${exporter.contentType} -> ${exporter.fileName()} ---`);
  console.log(exporter.exportAll(metrics));
  console.log();
}

const metrics = [
  new Metric('cpu', 0.93),
  new Metric('memory', 0.67),
  new Metric('disk', 0.41),
];

deliver(new CsvExporter(), metrics);
deliver(new JsonExporter(), metrics);
